src/NServiceBus.Core.Tests/StructConventionsTests.cs

```diff
@@ -0,0 +1,102 @@
+namespace NServiceBus.Core.Tests
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Runtime.InteropServices;
+    using System.Text;
+    using ApprovalTests;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class StructConventionsTests
+    {
+        [Test]
+        public void ApproveStructsWhichDontFollowStructGuidelines()
+        {
+            var approvalBuilder = new StringBuilder();
+            approvalBuilder.AppendLine(@"-------------------------------------------------- REMEMBER --------------------------------------------------
+CONSIDER defining a struct instead of a class if instances of the type are small and commonly short-lived or are commonly embedded in other objects.
+
+AVOID defining a struct unless the type has all of the following characteristics:
+   * It logically represents a single value, similar to primitive types(int, double, etc.).
+   * It has an instance size under 16 bytes.
+   * It is immutable.
+   * It will not have to be boxed frequently.
+
+In all other cases, you should define your types as classes.
+-------------------------------------------------- REMEMBER --------------------------------------------------
+");
+
+            var assembly = typeof(Endpoint).Assembly;
+            foreach (var type in assembly.GetTypes())
+            {
+                if (!type.IsValueType || type.IsEnum) continue;
+
+                var offendedRules = new List<string> { $"{type.FullName} violates the following rules:" };
+
+                InspectSizeOfStruct(type, offendedRules);
+                InspectMutabilityOfStruct(type, offendedRules);
+
+                if (offendedRules.Count <= 1) continue;
+                foreach (var offendedRule in offendedRules)
+                {
+                    approvalBuilder.AppendLine(offendedRule);
+                }
+                approvalBuilder.AppendLine();
+            }
+
+            Approvals.Verify(approvalBuilder.ToString());
+        }
+
+        static void InspectMutabilityOfStruct(Type type, List<string> offendedRules)
```

Comment:
It is a start. Then lets rename it

On Monday, July 25, 2016, Brandon Ording <notifications@github.com> wrote:

> In src/NServiceBus.Core.Tests/StructConventionsTests.cs
> <https://github.com/Particular/NServiceBus/pull/3963#discussion_r72095991>
> :
>
> > +
> > +                InspectSizeOfStruct(type, offendedRules);
> > +                InspectMutabilityOfStruct(type, offendedRules);
> > +
> > +                if (offendedRules.Count <= 1) continue;
> > +                foreach (var offendedRule in offendedRules)
> > +                {
> > +                    approvalBuilder.AppendLine(offendedRule);
> > +                }
> > +                approvalBuilder.AppendLine();
> > +            }
> > +
> > +            Approvals.Verify(approvalBuilder.ToString());
> > +        }
> > +
> > +        static void InspectMutabilityOfStruct(Type type, List<string> offendedRules)
>
> This appears to be misnamed, because it's actually looking to see if the
> struct has any reference-type members. Being a class doesn't inherently
> mean the type is mutable! Classes can be designed to be immutable as well.
>
> However. it seems like a true test of immutability would somewhat
> difficult to write. The easy things would be to check for the presence of
> public fields and properties with setters. I can't think of a way to verify
> that a method doesn't mutate state, though.
>
> Perhaps it would be better to change the naming of this to a "reference
> type checker" ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3963/files/59468bff365a248ecc8e1e1f2e97325fdbc7e01c#r72095991>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKoslhky-YfrK7kR3EA_lTpqELK3zLrks5qZOL2gaJpZM4JTWmG>
> .
>